### ISSUE TITLE
Task: NIRISS - SOSS

### ISSUE THREAD

**Issue by brechmos-stsci** [STScI-JWST/jwst#275]
To Do
- [x]  Reference file delivery to CRDS (INS)

**Comment by brechmos-stsci** [comment]
Starting work on this now.

**Comment by brechmos-stsci** [comment]
Craig's random notes:
- Look at the input FITS from Michael should be 3 planes
- One tabular model for each order and save inthe same asdf referenc efile, use keys of 1, 2, 3 and then the model 
- tabular model is in modeling

**Comment by brechmos-stsci** [comment]
I am using a copy of the data from @nden that was in her /user/dencheva/niriss-soss directory. The full frame data is:
```
f[0].data.shape
Out[6]: (3, 2048, 2048)
```
and when you plot the 3 "images":

![image](https://cloud.githubusercontent.com/assets/18348847/18716295/4a9fa5aa-7fea-11e6-8c59-d6f8a0cc0c6c.png)

**Comment by brechmos-stsci** [comment]
I created the `soss_wavelengths_configuration.asdf` from the  `soss_wavelengths_fullframe.fits` by using the `convert_polynomial_data.py`:
```
import sys
from astropy.io import fits
from astropy.modeling.models import Tabular2D
from asdf import AsdfFile

# Define the input and output files
input_fits = 'soss_wavelengths_fullframe.fits'
output_asdf = 'soss_wavelengths_configuration.asdf'

# Load the configuration data from the FITS file
f = fits.open(input_fits)

# Confirm the data looks correct
# The assumption for this is that the fits file has one HDU
# and the data in it is [3, 2048, 2048], well, at least 3 sets of data
if len(f) > 1 or not hasattr(f[0], 'data') or not f[0].data.shape[0] == 3:
    print('Data does not appear to be correct in {}'.format(input_fits))
    sys.exit()

# Copy over the reference data into Tabular classes
t1 = Tabular2D(lookup_table=f[0].data[0])
t2 = Tabular2D(lookup_table=f[0].data[1])
t3 = Tabular2D(lookup_table=f[0].data[2])

# Create the output ASDF file
fa = AsdfFile()

# Copy the data over the 1st order, 2nd order and 3rd order corrections
fa.tree[1] = t1
fa.tree[2] = t2
fa.tree[3] = t3

fa.write_to(output_asdf)
fa.close()
```

I have not confirmed the output file, but it appears to be approximately the correct size (i.e., similar to the input file).

**Comment by hbushouse** [comment]
@brechmos-stsci In your plot above of the 3 "images", are those the domain cutouts for each of the 3 spectral orders?

**Comment by brechmos-stsci** [comment]
@hbushouse They are the three spectral orders from the `soss_wavelengths_fullframe.fits` file
```
n [2]: f = fits.open('soss_wavelengths_fullframe.fits')
In [3]: [x.name for x in f]
Out[3]: ['PRIMARY']
In [4]: f[0].header
Out[4]:
SIMPLE  =                    T / conforms to FITS standard
BITPIX  =                  -32 / array data type
NAXIS   =                    3 / number of array dimensions
NAXIS1  =                 2048
NAXIS2  =                 2048
NAXIS3  =                    3
EXTEND  =                    T
In [6]: f[0].data.shape
Out[6]: (3, 2048, 2048)
```
That file has one HDU and the data in it is `(3, 2048, 2048)` and I plotted each of them.  

As @nden and I understand it, the soss-wavelengths_fullframe.fits file is to be used for the reference file (we are supposed to get an asdf reference file that contains the data, and this is the file to be used to create the reference file).

@nden please feel free to correct my wording...

**Comment by brechmos-stsci** [comment]
So.... The reference files (images above) have relevant data of size 2048 x 256.  The SOSS file I am "correcting", `soss_2A_results_int_001_2B.fits`, is apparently 96 x 2048:
```
```

**Comment by brechmos-stsci** [comment]
I sent some specific questions to Kevin Volk:

**Inputs**
My understand is that there will be 2 or 3 inputs for assigning the WCS:

1) Input Data: It will be a subarray (96 x 2048 or 256 x 2048).

2) Wavelength Correction File: It will be full frame (3x2048x2048, DMS orientation)

3) Rotation / Shift Correction of the input data


**The Questions**

So, there are a few things I need confirmed (early next week, if at all possible).

Q1) Are my assumptions about the input data in 1) above correct?

Q2) If there is the Rotation and Shift correction then I need some help understanding how they are applied. Is it shift and then rotate about the center?  What units are they in (degrees and pixels?)?  Which file will the be stored in and what are the key words?

Q3) What is the order of the whole process?  Will it be rotate/shift, correct wavelength?

Q4) There was talk, if I remember correctly, of the output coordinate system (same as input, v2/v3, sky etc).  I will need to know what the final output coordinate system is going to be in so the transforms can be added.

Q5) Are there any other corrections that need to happen or is it just what I have described above?

Q6) I have an example SOSS file, but could we get an input dataset, correction data, and corresponding output so we have something to compare against in order to test?

If you could get back to me early next week that would be great!

**Comment by brechmos-stsci** [comment]
Kevin Volk's answers (slightly reformatted)

**Inputs**

My understand is that there will be 2 or 3 inputs for assigning the WCS:

1) Input Data: It will be a subarray (96 x 2048 or 256 x 2048).

> It is also possible that a full frame SOSS measurement may be taken, which would be 2048x2048 pixels.  This is why the wavelength file is 2048x2048 pixels, even though the lower 1792 pixels will not have useful spectral data.  This would only be done if the target were relatively faint and needed the longer exposure per frame to increase the signal-to-noise ratio.


2) Wavelength Correction File: It will be full frame (3x2048x2048, DMS orientation)

> There are no values for the lower 2048x1792 pixels (x, y) in the wavelength file.  I think they are all zero  in that part.  There will also be a “standard” wavelength calibration spectrum that needs to be used, this is the spectrum from which the wavelength values in the reference file were obtained.


3) Rotation / Shift Correction of the input data


**The Questions**

So, there are a few things I need confirmed (early next week, if at all possible).

Q1) Are my assumptions about the input data in 1) above correct?

>See above.

Q2) If there is the Rotation and Shift correction then I need some help understanding how they are applied. Is it shift and then rotate about the center?  What units are they in (degrees and pixels?)?  Which file will the be stored in and what are the key words?

> The assumption is that one has a wavelength calibration observation from which the ramp slope image is available as well as the wavelength values for the orders.  The rotation and shift corrections are applied to the wavelength values/wavelength spectrum to match it to the science exposure (we do not want to rotate or shift the science spectra, these are left strictly alone).  So one has to (a) find the PWCPOS value for the science exposure in the header, (b) find the PWCPOS value for the wavelength standard spectrum, (c) rotate both the wavelength standard spectrum by the difference in angle (the PWCPOS values are in degrees), (d) cross-correlate the rotated wavelength calibration spectrum with the science spectrum to find the offsets (delta x, delta y) in pixels needed to match the trace of the wavelength standard to the trace of the science spectrum, and then (e) offset the standard wavelength values by these (delta x, delta y) pixel values with linear interpolation.  This results in a “new” wavelength grid that matches the science spectrum orientation and position for each order.

> Code for doing the cross-matching and the rotation were provided in the SOSS algorithm delivery.

Q3) What is the order of the whole process?  Will it be rotate/shift, correct wavelength?

> See above…rotate the wavelength reference files first, then cross-correlate the two spectra and shift the wavelength arrays accordingly.  No shifts or rotations are applied to the science data.

Q4) There was talk, if I remember correctly, of the output coordinate system (same as input, v2/v3, sky etc).  I will need to know what the final output coordinate system is going to be in so the transforms can be added.

> I am not sure about this.  The SOSS mode WCS is really just for getting the wavelengths assigned, and not for transforming back to the sky coordinates.  The source being observed is assumed to be a point source.  The distortions of the GR700XD prism make it difficult or impossible to relate the spectral pixels back to different sky positions, and we will have no way to measure the distortions involved as a function of position (as far as I know).  My first thought is to simply get the target (ra,dec) coordinates and assign these to every pixel.  However if this will cause issues in the pipeline then we have to think of something else.


Q5) Are there any other corrections that need to happen or is it just what I have described above?

> I think this is all we need.  In the pipeline there will be a background correction--which will not be available for the test data--but that is not of concern for testing the WCS step.

Q6) I have an example SOSS file, but could we get an input dataset, correction data, and corresponding output so we have something to compare against in order to test?

> We do not have these.  I have to get a “wavelength calibration spectrum” for you, and its associated PWCPOS value.  I should have this data around.  We then need to find some other SOSS observation to use to test the code, but it is likely to be at the same position and rotation (most of the SOSS data were taken in one block in the testing) so we may need to rotate/slightly shift some spectrum to give a test data set.

I am going to attempt to summarize and get confirmation...

**Comment by brechmos-stsci** [comment]
Input data in soss_2A_results_int_001_2B.fits for reference.

![image](https://cloud.githubusercontent.com/assets/18348847/18993323/97285360-86f0-11e6-8f07-56e6df4984ef.png)

**Comment by brechmos-stsci** [comment]
@nden @hbushouse 
The current set of code runs on the one set of data and one set of wavelength lookup files we have from Michael Fox.  The orientation of the data vs the orientation of the wavelength look up files does not appear to be consistent but we do not have any other data from them. 

I am going to create a PR for what is there, now, and we will have to make possible changes when we get more consistent data / wavelength correction files.

**Comment by hbushouse** [comment]
It's hard to compare to any old images or examples in docs, because I believe a lot of them were still in native detector orientation, instead of DMS, and I think it's also the case that they rotated the grism dispersion direction 90 deg in the instrument a year or so ago so that the entire spectrum would fall within a single detector readout amplifier (it used to spread across all 4 amps). So you have to be very careful to determine if what you're seeing is what we'll be getting out of the pipeline for current data.

**Comment by nden** [comment]
task implemented in #307 and reference file delivered by INS.
